client/tinychain/ml/linalg.py

```diff
@@ -273,3 +276,175 @@ def svd(matrix: Tensor) -> Tuple:
     """Return the singular value decomposition of the given `matrix`."""
 
     raise NotImplementedError("singular value decomposition")
+
+def sign_like(a, b):
+    return Number(If(b >= 0, a.abs(), -a.abs()))
+
+
+def dot(x1, x2):
+    return (x1 * x2).sum()
+
+@post_op
+def householder_bidiag(U: Tensor, W: Tensor, e: Tensor) -> Map:
+    """Householder's reduction to bidiagonal form"""
+
+    @post_op
+    def step(
+            cxt,
+            i: UInt,
+            n: UInt,
+            m: UInt,
+            U: Tensor,
+            W: Tensor,
+            e: Tensor,
+            scale: F32,
+            g: F32) -> Map:
+
+        @post_op
+        def update_cols(i: UInt, l: UInt, n: UInt, U: Tensor, scale: F32) -> F32:
+
+            @post_op
+            def step(j: UInt, n: UInt, i: UInt, U: Tensor, h: F32) -> Map:
+                return After(
+                    U[i:, j].write(U[i:, j] + U[i:, i] * dot(U[i:, i], U[i:, j]) / h),
+                    Map(j=j + 1, n=n, i=i, U=U, h=h)
+                )
+
+            @post_op
+            def cond(j: UInt, n: UInt) -> Bool:
+                return j < n
+
+            s, f = Tuple(After(U[i:, i].write(U[i:, i] / scale), [dot(U[i:, i], U[i:, i]), U[i, i]])).unpack(2)
```

Comment:
It would be more efficient to just slice `U` once rather than re-evaluating `U[i:, i]` repeatedly, e.g.
```python
def update_cols(cxt, ...):
    cxt.col = U[i:, i]
    # ...
    s, f = Tuple(After(cxt.col.write(cxt.col / scale), [dot(cxt.col, cxt.col)...
```